stackslib/src/net/download/nakamoto.rs

```diff
@@ -0,0 +1,4283 @@
+        self.state = NakamotoTenureDownloadState::GetTenureBlocks(
+            tenure_end_block.header.parent_block_id.clone(),
+        );
+        Ok(())
+    }
+
+    /// Determine how many blocks must be in this tenure.
+    /// Returns None if we don't have the start and end blocks yet.
+    pub fn tenure_length(&self) -> Option<u64> {
+        self.tenure_end_header
+            .as_ref()
+            .map(|(_hdr, tc_payload)| u64::from(tc_payload.previous_tenure_blocks))
+    }
+
+    /// Add downloaded tenure blocks to this machine.
+    /// If we have collected all tenure blocks, then return them and transition to the Done state.
+    ///
+    /// Returns Ok(Some([blocks])) if we got all the blocks in this tenure. The blocks will be in
+    /// ascending order by height, and will include the tenure-start block but exclude the
+    /// tenure-end block.
+    /// Returns Ok(None) if the given blocks were valid, but we still need more.  The pointer to
+    /// the next block to fetch (stored in self.state) will be updated.
+    /// Returns Err(..) if the blocks were invalid.
+    pub fn try_accept_tenure_blocks(
+        &mut self,
+        mut tenure_blocks: Vec<NakamotoBlock>,
+    ) -> Result<Option<Vec<NakamotoBlock>>, NetError> {
+        let NakamotoTenureDownloadState::GetTenureBlocks(block_cursor) = &self.state else {
+            warn!("Invalid state for this method";
+                  "state" => %self.state);
+            return Err(NetError::InvalidState);
+        };
+
+        if tenure_blocks.is_empty() {
+            // nothing to do
+            return Ok(None);
+        }
+
+        // blocks must be contiguous and in order from highest to lowest
+        let mut expected_block_id = block_cursor;
+        let mut count = 0;
+        for block in tenure_blocks.iter() {
+            if &block.header.block_id() != expected_block_id {
+                warn!("Unexpected Nakamoto block -- not part of tenure";
+                      "expected_block_id" => %expected_block_id,
+                      "block_id" => %block.header.block_id(),
+                      "state" => %self.state);
+                return Err(NetError::InvalidMessage);
+            }
+
+            let schnorr_signature = &block.header.signer_signature.0;
+            let message = block.header.signer_signature_hash().0;
+            if !schnorr_signature.verify(&self.start_aggregate_public_key, &message) {
+                warn!("Invalid block: bad signer signature";
+                      "tenure_id" => %self.tenure_id_consensus_hash,
+                      "block.header.block_id" => %block.header.block_id(),
+                      "start_aggregate_public_key" => %self.start_aggregate_public_key,
+                      "state" => %self.state);
+                return Err(NetError::InvalidMessage);
+            }
+
+            expected_block_id = &block.header.parent_block_id;
+            count += 1;
```

Comment:
I don't really understand what count is...? Is it block count? Could it be linked to the enumeration of the loop instead of keeping a sep mutable variable? I know you will still need to saturating add 1 to it may not be much cleaner

Reply:
The `TenureChange` payload indicates how many blocks were mined in the preceding tenure.  We're using `count` here to make sure that the remote node isn't serving us more blocks than expected (since we've already obtained the `TenureChange` for the tenure we're downloading).

Reply:
The `TenureChange` payload indicates how many blocks were mined in the preceding tenure.  We're using `count` here to make sure that the remote node isn't serving us more blocks than expected (since we've already obtained the `TenureChange` for the tenure we're downloading).